song-python-sdk/overture_song/model.py

```diff
@@ -0,0 +1,135 @@
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+# IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+#
+
+import logging
+
+from dataclasses import dataclass
+
+from overture_song import utils
+from overture_song.utils import SongClientException
+
+logging.basicConfig(level=logging.INFO)
+log = logging.getLogger("song.model")
+
+
+###############################
+# Utility Functions for Models
+###############################
+
+def get_optional_field(dic, field):
+    return dic[field] if field in dic else None
+
+
+def get_required_field(dic, field):
+    if field in dic:
+        return dic[field]
+    else:
+        raise SongClientException("study.id.dne", "The field '{}' does not exist in {}".format(field, str(dic)))
+
+
+################################
+#  Models
+################################
+
+
+class SongError(Exception):
+    FIELDS = ['stackTrace',
+              'errorId',
+              'httpStatusName',
+              'httpStatusCode',
+              'message',
+              'requestUrl',
+              'debugMessage',
+              'timestamp']
+
+    def __init__(self, data, response, debug=False):
+        self.stackTrace = data[SongError.FIELDS[0]]
+        self.errorId = data[SongError.FIELDS[1]]
+        self.httpStatusName = data[SongError.FIELDS[2]]
+        self.httpStatusCode = data[SongError.FIELDS[3]]
+        self.message = data[SongError.FIELDS[4]]
+        self.requestUrl = data[SongError.FIELDS[5]]
+        self.debugMessage = data[SongError.FIELDS[6]]
+        self.timestamp = data[SongError.FIELDS[7]]
+        self.response = response
+        self.__debug = debug
+
+    @classmethod
+    def is_song_error(cls, data):
+        for field in SongError.FIELDS:
+            if field not in data:
+                return False
+        return True
+
+    def __str__(self):
+        if self.__debug:
+            return utils.to_pretty_json_string(self.response.content)
+        else:
+            return "[SONG_SERVER_ERROR] {} @ {}: {}".format(self.errorId, self.timestamp, self.message)
+
+
+@dataclass(frozen=True)
+class ApiConfig(object):
+    server_url: str
+    study_id: str
+    access_token: str
+    debug: bool = False
+
+
+@dataclass(frozen=True)
+class ManifestEntry(object):
+    fileId: str
+    fileName: str
+    md5sum: str
+
+    @classmethod
+    def create_manifest_entry(cls, data):
+        d = data.__dict__
+        fileId = get_required_field(d, 'objectId')
+        fileName = get_required_field(d, 'fileName')
+        md5sum = get_required_field(d, 'fileMd5sum')
+        return ManifestEntry(fileId, fileName, md5sum)
+
+    def __str__(self):
+        return "{}\t{}\t{}".format(self.fileId, self.fileName, self.md5sum)
+
+
+class Manifest(object):
+    def __init__(self, analysis_id, manifest_entries=[]):
```

Comment:
```python
def __init__(self, analysis_id, manifest_entries=[]):
```
In here your have `manifest_entries` being assigned a mutable type as a default argument. 

The issue with this is that when python interprets this class/method definition it actually evaluates the default arguments and creates a list that is then shared across all calls to this method.

This causes the following bug in your case:

```python
first = Manifest("AN1")
first.add_entry("foo")
print(first)
# prints "foo" 

second = Manifest("AN2")
print(second)
# prints "foo" again because its sharing the same default instance as all invocations of __init__


second.add_entry("bar")
print(first)

# prints "foo" and "bar" for first Manifest even though we only added "bar" to the second Manifest

```

The way you avoid this bug is with the following: 
```python
    def __init__(self, analysis_id, manifest_entries=None):
        if manifest_entries is None:
            manifest_entries = []
```

This way the empty list is only evaluated and assigned at new instance creation.

Reply:
thank you! this probably explains a lot of odd behaviour i have had in the past